src/modules/site-v2/base/views/tools/genetic_mapping.py

```diff
@@ -22,45 +30,79 @@ def genetic_mapping():
   """
       This is the mapping submission page.
   """
-  title = 'Genetic Mapping'
-  alt_parent_breadcrumb = {"title": "Tools", "url": url_for('tools.tools')}
-  jwt_csrf_token = (get_jwt() or {}).get("csrf")
-  form = FileUploadForm()
-  # TODO: change
-  nemascan_container_url = 'https://github.com/AndersenLab/dockerfile/tree/nemarun/nemarun'
-  nemascan_github_url = 'https://github.com/AndersenLab/NemaScan'
-  return render_template('tools/genetic_mapping/mapping.html', **locals())
+  return render_template('tools/genetic_mapping/mapping.html', **{
+
+    # Page info
+    'title': 'Genetic Mapping',
+    'alt_parent_breadcrumb': {"title": "Tools", "url": url_for('tools.tools')},
+
+    # Form info
+    'jwt_csrf_token': (get_jwt() or {}).get("csrf"),
+    'form': FileUploadForm(),
+
+    # TODO: change
+    'nemascan_container_url': 'https://github.com/AndersenLab/dockerfile/tree/nemarun/nemarun',
+    'nemascan_github_url':    'https://github.com/AndersenLab/NemaScan',
+
+    # Species list
+    'species_list': SPECIES_LIST,
+    'species_fields': [
+      'name', 'short_name', 'project_num', 'wb_ver', 'latest_release',
+    ],
+  })
 
 
 @genetic_mapping_bp.route('/genetic-mapping/upload', methods = ['POST'])
 @jwt_required()
 def submit():
   form = FileUploadForm(request.form)
   user = get_current_user()
+
+  # If user is admin, allow them to bypass cache with URL variable
+  no_cache = bool(user_is_admin() and request.args.get("nocache", False))
+
+  # Validate form
   if not form.validate_on_submit():
     flash("You must include a description of your data and a TSV file to upload", "error")
     return redirect(url_for('genetic_mapping.mapping'))
-  
-  label = bleach.clean(request.form.get('label'))
-  props = {'label': label, 
-          'username': user.name,
-          'email': user.email,
-          'file': request.files['file']}          
+
+  # Read fields from form
+  label   = bleach.clean(request.form.get('label'))
+  species = bleach.clean(request.form.get('species'))
+
+  # Save uploaded file to server temporarily with unique generated name
+  # TODO: Is there a better way to generate this name? E.g. using a Tempfile?
+  local_path = os.path.join(uploads_dir, secure_filename(f'{ unique_id() }.tsv'))
+  request.files['file'].save(local_path)

```

Comment:
handle failed `save` operation and notify the user.

Reply:
Fixed in b1bae04850b89f5eb3dee32dc9fa474feb4b4698